examples/matrix.rs

```diff
@@ -1,125 +1,160 @@
+use multiversion::{multiversion, target::target_cfg_f};
+use rand::random;
+use slipstream::prelude::*;
 use std::fmt::Display;
+use std::hint::black_box;
 use std::iter;
-use std::num::Wrapping;
 use std::ops::Mul;
 use std::time::Instant;
 
-use multiversion::{multiversion, target};
-use rand::random;
-use slipstream::prelude::*;
+// Size of the matrices that are being computed
+//
+// This small matrix size is chosen such that the matrix fits in cache, which
+// means we don't have to implement cache blocking optimizations to achive
+// compute-bound performance and show the optimal effect of SIMD.
+//
+// The matrix size should be divisible by `V::LANES * CHUNK_VECS`.
+//
+const SIZE: usize = 80;
 
-const SIZE: usize = 1024;
-type V = wu32x8;
-type O = usizex8;
-const L: usize = V::LANES;
+// Number of output SIMD vectors we process together
+//
+// This should be greater than one for several reasons:
+// - Most compute-oriented CPUs can process multiple independent streams of
+//   arithmetic operations concurrently (e.g. current Intel and AMD CPUs can
+//   process two independent FMAs per CPU cycle). If we only feed those with a
+//   single stream of instructions that depend on each other, we lose perf.
+// - It is the granularity at which we amortize non-arithmetic operations like
+//   loop control code and scalar element broadcasting.
+//
+// Do not tune it too high or you will run out of CPU registers!
+//
+const CHUNK_VECS: usize = 10;
 
-#[derive(Debug, PartialEq)]
-struct Matrix(Vec<Wrapping<u32>>);
+// Vector type
+type V = f32x8;
 
-#[inline]
-fn at(x: usize, y: usize) -> usize {
-    y * SIZE + x
-}
+// Number of benchmark repetitions
+const RUNS: u32 = 10_000;
+
+// FIXME: Depending on how lucky you are with memory allocator lottery, you may
+//        or may not get a vector that's properly aligned for SIMD processing.
+//        Using a Vec<V> would be better from this perspective.
```

Comment:
I'm wondering about one thing. Allocate 7 more elements in the vector, look at the alignment of the first one and offset the start based on that. Then, if there's an assert at the start of the function, checking the proper alignment, the rest should be able to take advantage.

Reply:
In simple cases (think dot, not matmul), vectorize() could also take a hint from compiler autovectorizers and help here by splitting the loop into a (peel, body, tail) triplet, where body has an alignment and size that's a multiple of the SIMD vector size.

Reply:
...but sadly that won't work in the case where the user specifies multiple scalar slices with different alignment.